### PR TITLE
Resolve handler annotations under postponed evaluation

### PR DESCRIPTION
**Pull request by moonbox3**
## Summary
- Resolve handler message/context annotations via `typing.get_type_hints` so future-annotation strings are normalized to real types.
- Use resolved annotations in handler validation and explicit-type mode.
- Add regression tests for postponed annotations, unresolved forward refs, and invalid ctx annotations.

## Testing
- `pytest python/packages/core/tests/workflow/test_executor_future_annotations.py`

### CODE REVIEW

**Review by moonbox3 — COMMENTED** [review]
## Automated PR Review Round 1: CHANGES_REQUESTED

Issue: #1
The executor update to resolve postponed/forward-ref annotations via `typing.get_type_hints` is directionally correct, but the PR currently introduces failing/invalid tests and a risky CI workflow. The test suite changes must be fixed to reliably collect and to actually assert the intended regression protection. Additionally, the current `get_type_hints` exception fallback can silently bypass validation and should be made deterministic (fail fast or explicitly handle unresolved refs).

Blocking issues (3):
- Pytest collection likely fails: `python/packages/core/tests/workflow/test_executor_future_annotations.py` defines duplicate module-level tests that include an invalid `self` parameter (treated as a missing fixture) and overlaps with class-based tests; remove duplicates and/or fix signatures so tests reliably run.
- Validation may silently degrade: `_validate_handler_signature` catches `NameError`/`TypeError` from `typing.get_type_hints` and proceeds with raw `inspect` annotations, which can reintroduce string/forward-ref annotations and allow incorrect handler registration without a clear error; change behavior to deterministically fail or explicitly ensure annotations are fully resolved before continuing.
- CI/security risk: `.github/workflows/ai-triage.yml` embeds a secret token in a clone URL and executes unpinned external code/docker images at runtime, expanding supply-chain and credential exposure risk; must be removed or reworked to avoid token-in-URL and to pin/gate external execution appropriately.

Suggestions (4):
- Strengthen regression coverage by asserting the stored/validated context annotation is resolved (not a string/ForwardRef), e.g., check `get_origin(...) is WorkflowContext` and that type args are extracted as expected.
- Add negative tests for unresolved forward references (e.g., `WorkflowContext['MissingType']` or message `'DoesNotExist'`) and assert a clear, deterministic error path (rather than silent fallback).
- Consolidate overlapping test modules (`test_executor_future.py` vs `test_executor_future_annotations.py`) to reduce redundancy and confusion, keeping one focused file with both positive and negative cases.
- +1 more

File: python/packages/core/tests/workflow/test_executor_future_annotations.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        executor = ForwardRefCtxExecutor(id="forward_ref")
+        spec = executor._handler_specs[0]
+        assert spec["output_types"] == [int]
+        assert spec["workflow_output_types"] == [str]
+
+
+def test_executor_handler_future_annotations(self) -> None:
+    class FutureAnnotationExecutor(Executor):
+        @handler
+        async def handle(self, value: int, ctx: WorkflowContext[int]) -> None:
```
Comment (moonbox3) [inline python/packages/core/tests/workflow/test_executor_future_annotations.py:42]: Module-level tests below duplicate the class-based tests and include a `self` parameter (pytest will treat it as a fixture and fail). Remove duplicates and/or fix signatures (no `self`) so collection is reliable.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            async def handle(self, value: int, ctx: WorkflowContext[int, str]) -> None:
+                await ctx.send_message(value)
+                await ctx.yield_output("done")
+
+        executor = ForwardRefCtxExecutor(id="forward_ref")
+        spec = executor._handler_specs[0]
+        assert spec["output_types"] == [int]
+        assert spec["workflow_output_types"] == [str]
+
+
```
Comment (moonbox3) [inline python/packages/core/tests/workflow/test_executor_future_annotations.py:38]: These tests should assert postponed/forward-ref resolution (e.g., stored `ctx_annotation` is not a string/ForwardRef and `get_origin(...) is WorkflowContext`), otherwise they don’t strongly validate the intended fix.

File: python/packages/core/agent_framework/_workflows/_executor.py [hunk 3]
```diff
@@ -722,20 +719,28 @@ def _validate_handler_signature(
     if not skip_message_annotation and message_param.annotation == inspect.Parameter.empty:
         raise ValueError(f"Handler {func.__name__} must have a type annotation for the message parameter")
 
+    try:
+        type_hints = typing.get_type_hints(func, include_extras=True)
+    except (TypeError, NameError):
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_workflows/_executor.py:724]: Catching `TypeError`/`NameError` from `get_type_hints` and falling back to raw annotations can silently accept unresolved forward refs (reintroducing the original bug). Prefer failing fast with a handler/param-specific error or only allowing fallback when annotations are already concrete.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py (hunk 1 above)
Comment (moonbox3) [reply python/packages/core/tests/workflow/test_executor_future_annotations.py:42]: Automated status: queued for remediation in this pass.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py (hunk 2 above)
Comment (moonbox3) [reply python/packages/core/tests/workflow/test_executor_future_annotations.py:38]: Automated status: queued for remediation in this pass.

File: python/packages/core/agent_framework/_workflows/_executor.py (hunk 3 above)
Comment (moonbox3) [reply python/packages/core/agent_framework/_workflows/_executor.py:724]: Automated status: queued for remediation in this pass.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py (hunk 1 above)
Comment (moonbox3) [reply python/packages/core/tests/workflow/test_executor_future_annotations.py:42]: Automated status (r1): remediation attempted, but verification still failing (stage:code_quality, stage:unit_tests_subset); escalating.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py (hunk 2 above)
Comment (moonbox3) [reply python/packages/core/tests/workflow/test_executor_future_annotations.py:38]: Automated status (r1): remediation attempted, but verification still failing (stage:code_quality, stage:unit_tests_subset); escalating.

File: python/packages/core/agent_framework/_workflows/_executor.py (hunk 3 above)
Comment (moonbox3) [reply python/packages/core/agent_framework/_workflows/_executor.py:724]: Automated status (r1): remediation attempted, but verification still failing (stage:code_quality, stage:unit_tests_subset); escalating.

**Review by moonbox3 — COMMENTED** [review]
## Automated PR Review Round 2: CHANGES_REQUESTED

Issue: #1
The executor change to use `typing.get_type_hints()` for postponed/forward-ref annotations is directionally correct, but the PR is not safe to merge due to broken/duplicated tests and risky CI workflow additions. Pytest collection is likely to fail in the new future-annotations test module, and handler signature validation currently has non-deterministic fallback behavior when type-hint resolution fails. Additionally, the added GitHub Actions workflow patterns are non-hermetic and introduce concrete credential/supply-chain risks unrelated to the executor fix.

Blocking issues (3):
- Fix `python/packages/core/tests/workflow/test_executor_future_annotations.py`: remove duplicated test blocks and correct invalid module-level test signatures that take `self` (pytest will treat it as an unresolved fixture and fail collection or mis-collect).
- Remove the newly added GitHub Actions workflow changes from this PR or substantially harden them: avoid token-in-URL cloning patterns and pin any external container images (prefer digest) to prevent credential exposure and non-deterministic/supply-chain breakage; current behavior is unrelated to the executor fix and is merge-blocking.
- Make `_validate_handler_signature` deterministic when `get_type_hints()` fails: the current silent fallback to raw `inspect` annotations can reintroduce string/ForwardRef types and lead to inconsistent validation/registration; fail fast with a clear error or explicitly handle unresolved references rather than proceeding silently.

Suggestions (3):
- Add a negative test asserting unresolved forward references (e.g., `WorkflowContext['MissingType']` or decorator `input='MissingType'`) raise a clear, actionable error rather than being silently accepted.
- Tighten exception handling around `get_type_hints()` to only catch expected resolution errors and provide a helpful error message; consider passing appropriate `globalns/localns` (e.g., `func.__globals__`) to reduce accidental `NameError` in nested/local test types.
- Consolidate overlapping future-annotation tests into a single focused module and ensure at least one assertion verifies the resolved annotations are actual types (not `str`/`ForwardRef`) to demonstrate the regression fix.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        class ForwardRefCtxExecutor(Executor):
+            @handler
+            async def handle(self, value: int, ctx: WorkflowContext[int, str]) -> None:
+                await ctx.send_message(value)
+                await ctx.yield_output("done")
+
+        executor = ForwardRefCtxExecutor(id="forward_ref")
+        spec = executor._handler_specs[0]
+        assert spec["output_types"] == [int]
+        assert spec["workflow_output_types"] == [str]
```
Comment (moonbox3) [inline python/packages/core/tests/workflow/test_executor_future_annotations.py:36]: Module-level pytest tests below appear duplicated and include a `self` parameter; pytest will treat `self` as a fixture and fail collection/mis-collect. Remove the duplicate block or fix signatures and keep a single set of tests (class-based or module-level).

File: python/packages/core/tests/workflow/test_executor_future_annotations.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+
+def test_executor_handler_future_annotations(self) -> None:
+    class FutureAnnotationExecutor(Executor):
+        @handler
```
Comment (moonbox3) [inline python/packages/core/tests/workflow/test_executor_future_annotations.py:41]: This test intends to validate forward-ref resolution, but the annotation shown doesn’t appear quoted; consider using a quoted forward ref (or a locally defined class under postponed evaluation) and assert the resolved type is not a `str`/`ForwardRef`.

File: python/packages/core/agent_framework/_workflows/_executor.py [hunk 6]
```diff
@@ -722,20 +719,28 @@ def _validate_handler_signature(
...
+        type_hints = {}
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_workflows/_executor.py:725]: Catching `get_type_hints` failures and silently falling back to raw `inspect` annotations can reintroduce unresolved string/ForwardRef types and make handler validation non-deterministic. Prefer failing fast with a clear error or explicitly handling unresolved refs so registration doesn’t proceed with unresolved annotations.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py (hunk 4 above)
Comment (moonbox3) [reply python/packages/core/tests/workflow/test_executor_future_annotations.py:36]: Automated status: queued for remediation in this pass.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py (hunk 5 above)
Comment (moonbox3) [reply python/packages/core/tests/workflow/test_executor_future_annotations.py:41]: Automated status: queued for remediation in this pass.

File: python/packages/core/agent_framework/_workflows/_executor.py (hunk 6 above)
Comment (moonbox3) [reply python/packages/core/agent_framework/_workflows/_executor.py:725]: Automated status: queued for remediation in this pass.
